src/mach.rs

```diff
@@ -842,32 +843,30 @@ fn build_relocations(segment: &mut SegmentBuilder, artifact: &Artifact, symtab:
                         panic!("invalid DebugSection link")
                     }
 
+                    // from/to custom section
                     (Decl::Defined(DefinedDecl::Section(_)), _)
                     | (_, Decl::Defined(DefinedDecl::Section(_))) => {
                         panic!("relocations are not yet supported for custom sections")
                     }
-                    // various static function pointers in the .data section
-                    (
-                        Decl::Defined(DefinedDecl::Data { .. }),
-                        Decl::Defined(DefinedDecl::Function { .. }),
-                    ) => (true, X86_64_RELOC_UNSIGNED),
-                    (
-                        Decl::Defined(DefinedDecl::Data { .. }),
-                        Decl::Import(ImportKind::Function { .. }),
-                    ) => (true, X86_64_RELOC_UNSIGNED),
-                    // anything else is just a regular relocation/callq
-                    (_, Decl::Defined(DefinedDecl::Function { .. })) => {
-                        (false, X86_64_RELOC_BRANCH)
+
+                    // from data object
+                    (Decl::Defined(DefinedDecl::Data { .. }), _) => (true, X86_64_RELOC_UNSIGNED),
+
+                    // from function
+                    (Decl::Defined(DefinedDecl::Function { .. }), to) => match to {
+                        Decl::Defined(DefinedDecl::Function { .. }) => (false, X86_64_RELOC_BRANCH),
+                        Decl::Import(ImportKind::Function) => (false, X86_64_RELOC_BRANCH),
+
+                        Decl::Defined(DefinedDecl::Data { .. }) => (false, X86_64_RELOC_SIGNED),
+                        Decl::Import(ImportKind::Data) => (false, X86_64_RELOC_GOT_LOAD),
+
+                        // handled above
+                        Decl::Defined(DefinedDecl::Section { .. }) => unreachable!(),
+                    },
+
+                    (Decl::Import(_), _) => {
+                        unreachable!("Tried to relocate import???");
```

Comment:
Shouldn’t this be a panic, since it could technically be reachable, and is internal program error  ?

If not can we propagate an error up to user that they did something wrong, though I think artifact should have prevented this, which means it’s a programmer error and we should panic

Reply:
It is unreachable, as `Artifact::link` already errors out in this case:

https://github.com/m4b/faerie/blob/c4c81cac1d15fd86e66198b402353396876f9ee1/src/artifact.rs#L466-L468